File: Swift/AoC 2022/AoC 2022/Solutions/Day14.swift
//
//  Day14.swift
//  AoC 2022
//
//  Created by Simon Biickert on 2022-12-14.
//

import Foundation

class Day14: AoCSolution {
	let SAND_POINT = AoCCoord2D(x: 500, y: 0)
	let FALL_OFFSETS = [AoCCoord2D(x:  0, y: 1), // Down
						AoCCoord2D(x: -1, y: 1), // Down, left
						AoCCoord2D(x:  1, y: 1)] // Down, right
	
	override init() {
		super.init()
		day = 14
		name = "Regolith Reservoir"
	}
	
	override func solve(filename: String, index: Int) -> AoCResult {
		super.solve(filename: filename, index: index)
		
		let input = AoCUtil.readInputFile(named: filename, removingEmptyLines: false)
		let part1 = solvePartOne(input)
		print("Part One: there were \(part1) units of sand.")
		let part2 = solvePartTwo(input)
		print("Part Two: there were \(part2) units of sand.")

		return AoCResult(part1: String(part1), part2: String(part2))
	}
	
	private func solvePartOne(_ input: [String]) -> Int {
		let grid = parseGrid(input)
		let ext = grid.extent
		
		var sandCount = 0
		
		while (true) {
			let restPoint = dropSand(in: grid)
			if ext.contains(restPoint) {
				sandCount += 1
			}
			else {
				break
			}
		}
		
		return sandCount
	}
	
	private func solvePartTwo(_ input: [String]) -> Int {
		let grid = parseGrid(input)
		var ext = grid.extent
		
		let yMax = ext.max.y + 2
		drawLine(start: AoCCoord2D(x: 490-yMax, y: yMax), end: AoCCoord2D(x: 510+yMax, y: yMax), in: grid)
		ext = grid.extent
		
		var sandCount = 0
		
		while (true) {
			let restPoint = dropSand(in: grid)
			if restPoint == SAND_POINT {
				sandCount += 1
				break
			}
			else {
				sandCount += 1
			}
			//if (sandCount % 1000 == 0) { print ("\(sandCount)")}
		}
		
		return sandCount
	}

	private func dropSand(in grid: AoCGrid2D) -> AoCCoord2D {
		let ext = grid.extent
		var point = SAND_POINT
		while (ext.contains(point)) {
			var bMoved = false;
			for offset in FALL_OFFSETS {
				let checkPoint = point + offset
				if (grid.value(at: checkPoint) == grid.defaultValue) {
					point = checkPoint
					bMoved = true;
					break
				}
			}
			if (bMoved == false) {
				// Came to rest
				grid.setValue("o", at: point)
				return point
			}
		}
		return point
	}
	
	private func parseGrid(_ input: [String]) -> AoCGrid2D {
		let grid = AoCGrid2D(defaultValue: ".")
		
		for line in input {
			let pointDefs = line.split(separator: " -> ").map { String($0) }
			var points = [AoCCoord2D]()
			for pd in pointDefs {
				let xy = pd.split(separator: ",").map { Int($0)! }
				points.append(AoCCoord2D(x: xy.first!, y: xy.last!))
			}
			for c in 1..<points.count {
				drawLine(start: points[c-1], end: points[c], in: grid)
			}
		}
		
		grid.setValue("+", at: SAND_POINT)
		
		return grid
	}
	
	private func drawLine(start: AoCCoord2D, end: AoCCoord2D, in grid: AoCGrid2D) {
		//print("drawLine from \(start) to \(end)")
		let delta = end - start
		let md = start.manhattanDistance(to: end)
		let offset = AoCCoord2D(x: delta.x/md, y: delta.y/md)
		
		var drawCoord = start
		grid.setValue("#", at: drawCoord)
		//print("\t \(drawCoord)")
		for _ in 0..<md {
			drawCoord = drawCoord + offset
			grid.setValue("#", at: drawCoord)
			//print("\t \(drawCoord)")
		}
	}
}

File: Perl/day14.pl
#!/usr/bin/env perl
BEGIN {
    use Cwd;
    our $directory = cwd;
}

use lib $directory;

use Modern::Perl 2022;
use autodie;
use Data::Dumper;
#use Storable 'dclone';

use AOC::Util;
use AOC::SpatialUtil;

my $INPUT_PATH = '../input';
#my $INPUT_FILE = 'day14_test.txt';
my $INPUT_FILE = 'day14_challenge.txt';
our @input = read_input("$INPUT_PATH/$INPUT_FILE");

say "Advent of Code 2022, Day 14: Regolith Reservoir";

our $sand_pt = [500,0];
our @fall_offsets = ([0,1], [-1,1], [1,1]);

solve_part_one();
solve_part_two();

exit( 0 );

sub solve_part_one {
	my $grid = create_grid(@input);
	my $ext = G2D_extent($grid);
	
	my $sand_count = 0;
	while (1) {
		my $rest_point = drop_sand($grid);
		if (E2D_contains($ext, $rest_point)) {
			$sand_count++;
		}
		else {
			last;
		}
	}
	
	#G2D_print($grid);
	say "Part One: the amount of sand is $sand_count.";
}

sub solve_part_two {
	my $grid = create_grid(@input);
	my $ext = G2D_extent($grid);

	my $y_max = E2D_max( G2D_extent($grid) )->[1] + 2;
	draw_line($grid, [490 - $y_max, $y_max], [510 + $y_max, $y_max]);

	my $sand_count = 0;
	while (1) {
		my $rest_point = drop_sand($grid);
		if (C2D_equal($sand_pt, $rest_point)) {
			$sand_count++;
			last;
		}
		else {
			$sand_count++;
		}
		(E2D_contains($ext, $rest_point)) or die C2D_to_str($rest_point);

	}

	#G2D_print($grid);
	say "Part Two: the amount of sand is $sand_count.";
}

sub create_grid {
	my @input = @_;
	my $grid = G2D_create('.', 'queen');
	
	for my $line (@input) {
		my @coords = ();
		for my $coord (split(' -> ', $line)) {
			my @xy = split(',', $coord);
			push(@coords, \@xy);
		}
		for my $c (0..$#coords-1) {
			draw_line($grid, $coords[$c], $coords[$c+1]);
		}
	}
	G2D_set($grid, $sand_pt, '+');
	#G2D_print($grid);
	return $grid;
}

sub draw_line {
	my ($grid, $start, $end) = @_;
	#say C2D_to_str($start) . ' to ' . C2D_to_str($end);
	my $delta = C2D_delta($start, $end);
	my $md = C2D_manhattan($start, $end);
	my $offset = [$delta->[0]/$md, $delta->[1]/$md];
	
	my $draw_coord = $start;
	G2D_set($grid, $draw_coord, '#');
	#say "\tDrawing " . C2D_to_str($draw_coord);
	for my $i (0..$md-1) {
		$draw_coord = C2D_add($draw_coord, $offset);
		G2D_set($grid, $draw_coord, '#');
		#say "\tDrawing " . C2D_to_str($draw_coord);
	}
}

sub drop_sand {
	my $grid = shift;
	my $ext = G2D_extent($grid);
	my $point = C2D_add($sand_pt, [0,0]);
	while (E2D_contains($ext, $point)) {
		my $b_moved = 0;
		for my $offset (@fall_offsets) {
			my $check = C2D_add($point, $offset);
			if (G2D_get($grid, $check) eq '.') {
				$point = $check;
				$b_moved = 1;
				last;
			}
		}
		if (!$b_moved) {
			# Came to rest
			G2D_set($grid, $point, 'o');
			return $point;
		}
	}
	return $point;
}